Figure out MacOS compilation

Continuation of #42.

There seems to be a few things we need to fix up:
- [ ] deprecated Mac OS X API Calls
- [ ] struct vs class mismatches - /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/weapon_xml.h:11:1: warning: 'weapon_info' defined as a struct here but previously declared as a class; this is valid, but may result in linker errors under the Microsoft C++ ABI [-Wmismatched-tags]
- [ ] Bad GCC warnings on CLang - /usr/local/include/boost/proto/expr.hpp:140:44: warning: unknown warning group '-Wdeprecated-copy', ignored [-Wunknown-warning-option]
- [ ] OpenAL seems to be deprecated on Mac OSX 10.15, they prefer using AVAudioEngine.

I've had a look at the CI build, and tried to condense it down to just unique errors.
I have no idea what I'm doing, so sorry if this is not helpful...

```
####### Error 1 ########
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Ice/IceRandom.cpp:12:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Ice/Stdafx.h:1:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Stdafx.h:44:
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/opcodetypes.h:176:15: error: typedef redefinition with different types ('int64' (aka 'long long') vs 'long')
typedef int64 intmax_t;
              ^
/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/_types/_intmax_t.h:32:25: note: previous definition is here
typedef __INTMAX_TYPE__ intmax_t;
                        ^
####### Error 2 ########
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Ice/IceAABB.cpp:21:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Ice/Stdafx.h:1:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Stdafx.h:44:
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/opcodetypes.h:178:16: error: typedef redefinition with different types ('uint64' (aka 'unsigned long long') vs 'unsigned long')
typedef uint64 uintmax_t;
               ^
/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/_types/_uintmax_t.h:32:26: note: previous definition is here
typedef __UINTMAX_TYPE__ uintmax_t;
                         ^

####### Error 3 ########
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/OPC_HybridModel.cpp:85:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/Stdafx.h:45:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/cmd/collide2/opcodesysdef.h:50:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/gfx/quaternion.h:4:
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/gfx/vec.h:6:
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/endianness.h:102:14: warning: 'NXSwapHostLongToLittle' is deprecated [-Wdeprecated-declarations]
    l.i[0] = le32_to_cpu( l.i[1] );
             ^
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/endianness.h:27: error: typedef redefinition with different types ('int64' (aka 'long long') vs 'long')
:33: note: expanded from macro 'le32_to_cpu'
typedef int64 intmax_t;
              ^
/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/_types/_intmax_t.h:32:25: note: previous definition is here
    # define le32_to_cpu( x ) ( NXSwapHostLongToLittle( x ) )
                                ^
/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/architecture/byte_order.h:345:34: note: 'NXSwapHostLongToLittle' has been explicitly marked deprecated here
static __inline__ __attribute__((deprecated))
                                 ^
typedef __INTMAX_TYPE__ intmax_t;
                        ^

####### Error 4 ########
In file included from /Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/gfx/mesh.cpp:48:
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/gfx/sphere.h:86:18: warning: 'SphereMesh::ProcessDrawQueue' hides overloaded virtual function [-Woverloaded-virtual]
    virtual void ProcessDrawQueue( int whichpass, int which, bool zsort, const QVector &sortctr );
                 ^
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/gfx/mesh.h:344:18: note: hidden overloaded virtual function 'Mesh::ProcessDrawQueue' declared here: type mismatch at 1st parameter ('size_t' (aka 'unsigned long') vs 'int')
    virtual void ProcessDrawQueue( size_t whichpass, int whichdrawqueue, bool zsort, const QVector &sortctr );
                 ^
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/vsfilesystem.cpp:757:19: error: no matching function for call to 'scandir'
        int ret = scandir( selectcurrentdir.c_str(), &dirlist, selectdirs, 0 );
                  ^~~~~~~
/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/dirent.h:149:5: note: candidate function not viable: no known conversion from 'int (struct dirent *)' to 'int (*)(const struct dirent *)' for 3rd argument
int scandir(const char *, struct dirent ***,
    ^
/Users/runner/work/Vega-Strike-Engine-Source/Vega-Strike-Engine-Source/engine/src/vsfilesystem.cpp:771:15: error: no matching function for call to 'scandir'
    int ret = scandir( selectcurrentdir.c_str(), &dirlist, selectdirs, 0 );
              ^~~~~~~
/Applications/Xcode_12.4.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk/usr/include/dirent.h:149:5: note: candidate function not viable: no known conversion from 'int (struct dirent *)' to 'int (*)(const struct dirent *)' for 3rd argument
int scandir(const char *, struct dirent ***,
    ^
```

--- a/.github/workflows/Windows-CI.yml
+++ b/.github/workflows/Windows-CI.yml
@@ -26,7 +26,7 @@ jobs:
     steps:
       # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
       - name: checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
 
       - name: install-cmake
         uses: lukka/get-cmake@v3.21.2

--- a/.github/workflows/Windows-CI.yml
+++ b/.github/workflows/Windows-CI.yml
@@ -26,7 +26,7 @@ jobs:
     steps:
       # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
       - name: checkout
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 #v3.5.3
 
       - name: install-cmake
         uses: lukka/get-cmake@v3.21.2

--- a/.github/workflows/codeql.yml
+++ b/.github/workflows/codeql.yml
@@ -16,28 +16,20 @@ jobs:
 
     strategy:
       fail-fast: false
-      # matrix:
-        # Override automatic language detection by changing the below list
-        # Supported options are ['csharp', 'cpp', 'go', 'java', 'javascript', 'python']
-        # language: ['cpp', 'python']
-        # Learn more...
-        # https://docs.github.com/en/github/finding-security-vulnerabilities-and-errors-in-your-code/configuring-code-scanning#overriding-automatic-language-detection
+      matrix:
+        language: ['cpp', 'python']
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@v2
+      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
       with:
         fetch-depth: 2
 
     # Initializes the CodeQL tools for scanning.
     - name: Initialize CodeQL
-      uses: github/codeql-action/init@v1
-      # with:
-        # languages: ${{ matrix.language }}
-        # If you wish to specify custom queries, you can do so here or in a config file.
-        # By default, queries listed here will override any specified in a config file.
-        # Prefix the list here with "+" to use these queries and those in the config file.
-        # queries: ./path/to/local/query, your-org/your-repo/queries@main
+      uses: github/codeql-action/init@0225834cc549ee0ca93cb085b92954821a145866 #v2.3.5
+      with:
+        languages: ${{ matrix.language }}
 
     - name: Bootstrap and Build the Code
       run: |
@@ -45,4 +37,4 @@ jobs:
         script/build.sh
 
     - name: Perform CodeQL Analysis
-      uses: github/codeql-action/analyze@v1
+      uses: github/codeql-action/analyze@0225834cc549ee0ca93cb085b92954821a145866 #v2.3.5

--- a/.github/workflows/codeql.yml
+++ b/.github/workflows/codeql.yml
@@ -21,7 +21,7 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+      uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 #v3.5.3
       with:
         fetch-depth: 2
 

--- a/.github/workflows/fortify-on-demand-scan.yml
+++ b/.github/workflows/fortify-on-demand-scan.yml
@@ -16,7 +16,7 @@ jobs:
     # Steps represent a sequence of tasks that will be executed as part of the job
     steps:
       # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
-      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+      - uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 #v3.5.3
 
       - name: Fortify on Demand Scan
         # You may pin to the exact commit or the version.

--- a/.github/workflows/fortify-on-demand-scan.yml
+++ b/.github/workflows/fortify-on-demand-scan.yml
@@ -16,7 +16,7 @@ jobs:
     # Steps represent a sequence of tasks that will be executed as part of the job
     steps:
       # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
-      - uses: actions/checkout@v2
+      - uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
 
       - name: Fortify on Demand Scan
         # You may pin to the exact commit or the version.

--- a/.github/workflows/gh-actions-pr.yml
+++ b/.github/workflows/gh-actions-pr.yml
@@ -169,10 +169,8 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@v2
+      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
       with:
-        # We must fetch at least the immediate parents so that if this is
-        # a pull request then we can checkout the head.
         fetch-depth: 2
         submodules: false
 

--- a/.github/workflows/gh-actions-pr.yml
+++ b/.github/workflows/gh-actions-pr.yml
@@ -169,7 +169,7 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+      uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 #v3.5.3
       with:
         fetch-depth: 2
         submodules: false

--- a/.github/workflows/gh-actions-release.yml
+++ b/.github/workflows/gh-actions-release.yml
@@ -134,15 +134,10 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@v2
+      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
       with:
-        # We must fetch at least the immediate parents so that if this is
-        # a pull request then we can checkout the head.
         fetch-depth: 2
         submodules: false
-
-    # where in a pull request we want to move to the head of the PR and not be at the
-    # merge commit; here we want to be at the HEAD of the ref
 
     - name: Extract tag name and short SHA
       shell: bash

--- a/.github/workflows/gh-actions-release.yml
+++ b/.github/workflows/gh-actions-release.yml
@@ -134,7 +134,7 @@ jobs:
 
     steps:
     - name: Checkout repository
-      uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+      uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 #v3.5.3
       with:
         fetch-depth: 2
         submodules: false

--- a/.github/workflows/macos-ci.yml
+++ b/.github/workflows/macos-ci.yml
@@ -1,0 +1,98 @@
+name: 'MacOS-CI'
+
+on:
+  push:
+  pull_request:
+
+  # Allows you to run this workflow manually from the Actions tab
+  workflow_dispatch:
+
+jobs:
+  build:
+    name: Build
+    runs-on: ${{ matrix.os }}
+    continue-on-error: true
+
+    defaults:
+      run:
+        shell: bash
+
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - macos-11
+          - macos-12
+        compiler:
+          - clang
+          - gcc
+        homebrew-gl:
+          - true
+          # - false
+        homebrew-al:
+          - true
+          - false
+
+    steps:
+
+      # The following dependencies are already present within macos-* images:
+      # - clang (llvm)
+      # - cmake
+      # - expat
+      # - gcc
+      # - git
+      # - jpeg
+      # - libpng
+      # - libvorbis
+      # - python
+      - name: Install dependencies using homebrew
+        run: brew install boost-python3 gtk+3 gtkglext sdl
+
+      # The following Apple-provided libraries are deprecated:
+      # * OpenGL as of MacOS 10.14
+      # * GLUT as of MacOS 10.9
+      - name: Optionally install homebrewed OpenGL and GLUT
+        if: ${{ matrix.homebrew-gl }}
+        run: |
+          brew install mesa mesa-glu freeglut
+          ln -s /usr/local/include/GL /usr/local/include/OpenGL
+          ln -s /usr/local/include/GL /usr/local/include/GLUT
+        # ln -s /usr/local/lib/libGL.dylib /usr/local/lib/libOpenGL.dylib
+        # find /usr/local/lib/ -iname '*gl*.dylib'
+
+      # The Apple-provided OpenAL is deprecated as of MacOS 10.15
+      - name: Optionally install homebrewed OpenAL
+        if: ${{ matrix.homebrew-al }}
+        run: brew install openal-soft
+
+      - name: Check out repository
+        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+        with:
+          fetch-depth: 2
+          submodules: false
+
+      - name: Conditionally relocate to PR HEAD
+        if: github.event.pull_request
+        run: git checkout HEAD^2
+
+      - name: Build it
+        env:
+          MY_OS_NAME: macos
+          COMPILER:   ${{ matrix.compiler }}
+          FLAGS:      -DCMAKE_FIND_FRAMEWORK=LAST
+          OPENALDIR:  "/usr/local/opt/openal-soft"
+        run: script/cibuild $FLAGS
+
+#      - name: Test
+#        working-directory: ${{github.workspace}}/build
+#        env:
+#          GTEST_OUTPUT: xml
+#          GTEST_COLOR: 1
+#        run: ctest -V
+#
+#      - name: Upload test results
+#        uses: actions/upload-artifact@0b7f8abb1508181956e8e162db84b466c27e18ce #v3.1.2
+#        if: failure()
+#        with:
+#          name: test_results_xml
+#          path: ${{github.workspace}}/build/test-results/**/*.xml

--- a/.github/workflows/macos-ci.yml
+++ b/.github/workflows/macos-ci.yml
@@ -66,7 +66,7 @@ jobs:
         run: brew install openal-soft
 
       - name: Check out repository
-        uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab #v3.5.2
+        uses: actions/checkout@c85c95e3d7251135ab7dc9ce3241c5835cc595a9 #v3.5.3
         with:
           fetch-depth: 2
           submodules: false

--- a/.gitignore
+++ b/.gitignore
@@ -1,6 +1,8 @@
 bin/
 build/
 build-*/
+cmake-build/
+cmake-build-*/
 data/
 ext/
 *.pyc

--- a/engine/src/cmd/basecomputer.cpp
+++ b/engine/src/cmd/basecomputer.cpp
@@ -1,11 +1,10 @@
 // -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil -*-
 
-/**
+/*
  * basecomputer.cpp
  *
- * Copyright (C) 2003 Mike Byron
- * Copyright (C) 2019-2020 Stephen G. Tuggy, pyramid3d, and other Vega Strike
- * contributors
+ * Copyright (C) 2001-2023 Daniel Horn, Mike Byron, pyramid3d,
+ * Stephen G. Tuggy, and other Vega Strike contributors.
  *
  * https://github.com/vegastrike/Vega-Strike-Engine-Source
  *
@@ -2842,23 +2841,22 @@ void BaseComputer::loadNewsControls( void )
     recalcTitle();
 }
 
-#if defined (__APPLE__)
-static int nodirs( struct dirent *entry )
-#else
 static int nodirs( const struct dirent * entry )
-#endif
 {
 #if defined (_WIN32) || defined(__HAIKU__)
     //Have to check if we have the full path or just relative (which would be a problem)
-    std::string tmp = VSFileSystem::homedir+"/save/"+entry->d_name;
+    std::string tmp = VSFileSystem::homedir + "/save/" + entry->d_name;
     struct stat s;
-    if (stat( tmp.c_str(), &s ) < 0)
+    if (stat( tmp.c_str(), &s ) < 0) {
         return string( entry->d_name ) != "." && string( entry->d_name ) != "..";
-    if ( (s.st_mode&S_IFDIR) == 0 && string( entry->d_name ) != "." && string( entry->d_name ) != ".." )
+    }
+    if ( (s.st_mode & S_IFDIR) == 0 && string( entry->d_name ) != "." && string( entry->d_name ) != ".." ) {
         return 1;
+    }
 #else
-    if (entry->d_type != DT_DIR && string( entry->d_name ) != "." && string( entry->d_name ) != "..")
+    if (entry->d_type != DT_DIR && string( entry->d_name ) != "." && string( entry->d_name ) != "..") {
         return 1;
+    }
 #endif
     return 0;
 }
@@ -2877,11 +2875,7 @@ static int datesort( const void *v1, const void *v2 )
     return s1.st_mtime-s2.st_mtime;
 }
 
-#if (defined (__FREEBSD__)) || (defined (_WIN32) && !defined (__CYGWIN__ ) ) || (defined (__GLIBC_MINOR__) && __GLIBC_MINOR__ >= 10) || defined(__HAIKU__)
 typedef int (*scancompare)( const struct dirent **v1, const struct dirent **v2 );
-#else
-typedef int (*scancompare)( const void *v1, const void *v2 );
-#endif
 
 //Load the controls for the News display.
 void BaseComputer::loadLoadSaveControls( void )

--- a/engine/src/cmd/collide2/opcodetypes.h
+++ b/engine/src/cmd/collide2/opcodetypes.h
@@ -171,7 +171,7 @@ typedef int ptrdiff_t;
 #define _PTRDIFF_T_DEFINED
 #endif
 
-#if !defined(CS_HAVE_INTMAX_T) && !__APPLE__
+#if !defined(CS_HAVE_INTMAX_T) && !defined(__APPLE__)
 /// Greatest-width integer
 typedef int64 intmax_t;
 /// Greatest-width unsigned integer

--- a/engine/src/cmd/collide2/opcodetypes.h
+++ b/engine/src/cmd/collide2/opcodetypes.h
@@ -171,7 +171,7 @@ typedef int ptrdiff_t;
 #define _PTRDIFF_T_DEFINED
 #endif
 
-#if !defined(CS_HAVE_INTMAX_T)
+#if !defined(CS_HAVE_INTMAX_T) && !__APPLE__
 /// Greatest-width integer
 typedef int64 intmax_t;
 /// Greatest-width unsigned integer

--- a/engine/src/endianness.h
+++ b/engine/src/endianness.h
@@ -1,36 +1,35 @@
-/**
-* endianness.h
-*
-* Copyright (C) 2001-2002 Daniel Horn
-* Copyright (C) 2002-2019 pyramid3d and other Vega Strike Contributors
-* Copyright (C) 2019-2022 Stephen G. Tuggy and other Vega Strike Contributors
-*
-* https://github.com/vegastrike/Vega-Strike-Engine-Source
-*
-* This file is part of Vega Strike.
-*
-* Vega Strike is free software: you can redistribute it and/or modify
-* it under the terms of the GNU General Public License as published by
-* the Free Software Foundation, either version 2 of the License, or
-* (at your option) any later version.
-*
-* Vega Strike is distributed in the hope that it will be useful,
-* but WITHOUT ANY WARRANTY; without even the implied warranty of
-* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-* GNU General Public License for more details.
-*
-* You should have received a copy of the GNU General Public License
-* along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
-*/
+/*
+ * endianness.h
+ *
+ * Copyright (C) 2001-2023 Daniel Horn, pyramid3d, Stephen G. Tuggy
+ * and other Vega Strike Contributors
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
 
 #ifndef _ENDIANNESS_H
 #define _ENDIANNESS_H
 double DONTUSE__NXSwapBigDoubleToLittleEndian( double x );
 
 
-#if defined (__HAIKU__) //For unknow reasons, Haiku don't fit into any case below
+#if defined (__HAIKU__) //For unknown reasons, Haiku don't fit into any case below
     #include <endian.h>
-#elif defined (__APPLE__) || defined (MACOSX) || defined (BSD) || defined (__FreeBSD__)
+#elif defined (BSD) || defined (__FreeBSD__)
     #include <machine/endian.h>
 #else
 
@@ -45,17 +44,24 @@ double DONTUSE__NXSwapBigDoubleToLittleEndian( double x );
     # define __BYTE_ORDER 0
     #endif
 #endif
-#if defined (__APPLE__) || defined (MACOSX)
-    #include <machine/endian.h>
-    #if __BYTE_ORDER == __BIG_ENDIAN
-    # include <machine/byte_order.h>
-    # define le32_to_cpu( x ) ( NXSwapHostLongToLittle( x ) )
-    # define le16_to_cpu( x ) ( NXSwapHostShortToLittle( x ) )
-    # define le64_to_cpu( x ) ( DONTUSE__NXSwapBigDoubleToLittleEndian( x ) )
+
+#if defined (__APPLE__)
+    #if defined (__BIG_ENDIAN__)
+        #if defined (__x86_64__)
+            #include <libkern/OSByteOrder.h>
+            #define le16_to_cpu( x ) ( (unsigned long)      OSSwapLittleToHostInt16( (uint16_t) x ) )
+            #define le32_to_cpu( x ) ( (unsigned short)     OSSwapLittleToHostInt32( (uint32_t) x ) )
+            #define le64_to_cpu( x ) ( (unsigned long long) OSSwapLittleToHostInt64( (uint64_t) x ) )
+        #else
+            #include <architecture/byte_order.h>
+            #define le16_to_cpu( x ) NXSwapLittleShortToHost( x )
+            #define le32_to_cpu( x ) NXSwapLittleLongToHost( x )
+            #define le64_to_cpu( x ) NXSwapLittleLongLongToHost( x )
+        #endif
     #else
-    # define le32_to_cpu( x ) (x)
-    # define le16_to_cpu( x ) (x)
-    # define le64_to_cpu( x ) (x)
+        #define le32_to_cpu( x ) (x)
+        #define le16_to_cpu( x ) (x)
+        #define le64_to_cpu( x ) (x)
     #endif
 #else
     #if defined (IRIX) || (defined (__SVR4) && defined (__sun ) )

--- a/engine/src/gfx/aligned.h
+++ b/engine/src/gfx/aligned.h
@@ -1,27 +1,27 @@
-/**
-* aligned.h
-*
-* Copyright (C) 2001-2002 Daniel Horn
-* Copyright (C) 2002-2019 pyramid3d and other Vega Strike Contributors
-* Copyright (C) 2019-2022 Stephen G. Tuggy and other Vega Strike Contributors
-*
-* https://github.com/vegastrike/Vega-Strike-Engine-Source
-*
-* This file is part of Vega Strike.
-*
-* Vega Strike is free software: you can redistribute it and/or modify
-* it under the terms of the GNU General Public License as published by
-* the Free Software Foundation, either version 2 of the License, or
-* (at your option) any later version.
-*
-* Vega Strike is distributed in the hope that it will be useful,
-* but WITHOUT ANY WARRANTY; without even the implied warranty of
-* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-* GNU General Public License for more details.
-*
-* You should have received a copy of the GNU General Public License
-* along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
-*/
+/*
+ * aligned.h
+ *
+ * Copyright (C) 2001-2023 Daniel Horn, klaussfreire, pyramid3d,
+ * Stephen G. Tuggy, Benjamen R. Meyer, and other Vega Strike contributors.
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
+
 
 #ifndef __ALIGNED_H
 #define __ALIGNED_H
@@ -85,42 +85,60 @@ public:
     
     typename std::allocator<T>::pointer _align (typename std::allocator<void>::const_pointer p)
     {
-        if (p == 0)
-            return p;
-        
-        char *vrv;
+        if (p == nullptr) {
+            return static_cast<typename std::allocator<T>::pointer>(const_cast<typename std::allocator<void>::pointer>(p));
+        }
+
+        uint8_t *vrv;
         
         // Make room for the offset
-        vrv = (char*)((size_t*) p + 1);
+        vrv = (uint8_t*)((size_t*) p + 1);
         
         // Align
-        vrv += (ALIGN - (vrv - (char*)NULL) % ALIGN) % ALIGN;
+        vrv += (ALIGN - (vrv - (uint8_t*)nullptr) % ALIGN) % ALIGN;
         
         // Store offset
-        *((size_t*)vrv - 1) = vrv - (char*)p;
-        
-        return __alpn(vrv, ALIGN);
+        *((size_t*)vrv - 1) = vrv - (uint8_t*)p;
+
+        void *vrv2 = static_cast<void *>(vrv);
+        typename std::allocator<T>::pointer vrv3 = static_cast<typename std::allocator<T>::pointer>(vrv2);
+
+        return __alpn(vrv3, ALIGN);
     }
 
-    void _dealign (typename std::allocator<T>::pointer p)
+    typename std::allocator<T>::pointer _dealign (typename std::allocator<T>::pointer p)
     {
-        if (p == 0)
+        if (p == 0) {
             return p;
-        
+        }
+
+        // Convert our input to pointer-to-void as an intermediate step
+        void *p2 = static_cast<void *>(p);
+
         // De-align
-        char *vrv = (char*)p - *((size_t*)p - 1);
-        
-        return (typename std::allocator<T>::pointer)vrv;
+        uint8_t *vrv = (uint8_t*)p2 - *((size_t*)p2 - 1);
+
+        // Now convert back to pointer-to-void on the way out
+        void *vrv2 = static_cast<void *>(vrv);
+
+        // And back to the specific pointer type desired
+        typename std::allocator<T>::pointer vrv3 = static_cast<typename std::allocator<T>::pointer>(vrv2);
+
+        // And return
+        return vrv3;
     }
     
-    typename std::allocator<T>::pointer allocate (typename std::allocator<T>::size_type n, typename std::allocator<void>::const_pointer hint=0)
+    typename std::allocator<T>::pointer allocate (typename std::allocator<T>::size_type n, typename std::allocator<void>::const_pointer hint = 0)
     {
-        return __alpn(_align(std::allocator<T>::allocate(n+_OVERHEAD, _dealign(hint))), ALIGN);
+        typename std::allocator<void>::pointer hint2 = const_cast<typename std::allocator<void>::pointer>(hint);
+        typename std::allocator<T>::pointer    hint3 = static_cast<typename std::allocator<T>::pointer>(hint2);
+
+        return __alpn(_align(std::allocator<T>::allocate(n + _OVERHEAD, _dealign(hint3))), ALIGN);
     }
     
     void deallocate (typename std::allocator<T>::pointer p, typename std::allocator<T>::size_type n)
     {
-        std::allocator<T>::deallocate(_dealign(p), n+_OVERHEAD);
+        std::allocator<T>::deallocate(_dealign(p), n + _OVERHEAD);
     }
 };
 

--- a/engine/src/gfx/jpeg_memory.h
+++ b/engine/src/gfx/jpeg_memory.h
@@ -1,27 +1,26 @@
-/**
-* jpeg_memory.h
-*
-* Copyright (C) 2001-2002 Daniel Horn
-* Copyright (C) 2002-2019 pyramid3d and other Vega Strike Contributors
-* Copyright (C) 2019-2022 Stephen G. Tuggy and other Vega Strike Contributors
-*
-* https://github.com/vegastrike/Vega-Strike-Engine-Source
-*
-* This file is part of Vega Strike.
-*
-* Vega Strike is free software: you can redistribute it and/or modify
-* it under the terms of the GNU General Public License as published by
-* the Free Software Foundation, either version 2 of the License, or
-* (at your option) any later version.
-*
-* Vega Strike is distributed in the hope that it will be useful,
-* but WITHOUT ANY WARRANTY; without even the implied warranty of
-* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-* GNU General Public License for more details.
-*
-* You should have received a copy of the GNU General Public License
-* along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
-*/
+/*
+ * jpeg_memory.h
+ *
+ * Copyright (C) 2001-2023 Daniel Horn, pyramid3d, Stephen G. Tuggy
+ * and other Vega Strike Contributors
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
 
 #include <stdio.h>
 #define JPEG_SUPPORT
@@ -59,7 +58,6 @@ typedef unsigned char boolean;
 extern "C" {
 //#define XMD_H
 #include <jconfig.h>
-#include <jmorecfg.h>
 #include <jpeglib.h>
 }
 /*--------------

--- a/engine/src/gfx/particle.h
+++ b/engine/src/gfx/particle.h
@@ -1,27 +1,26 @@
-/**
-* particle.h
-*
-* Copyright (C) 2001-2002 Daniel Horn
-* Copyright (C) 2002-2019 pyramid3d and other Vega Strike Contributors
-* Copyright (C) 2019-2022 Stephen G. Tuggy and other Vega Strike Contributors
-*
-* https://github.com/vegastrike/Vega-Strike-Engine-Source
-*
-* This file is part of Vega Strike.
-*
-* Vega Strike is free software: you can redistribute it and/or modify
-* it under the terms of the GNU General Public License as published by
-* the Free Software Foundation, either version 2 of the License, or
-* (at your option) any later version.
-*
-* Vega Strike is distributed in the hope that it will be useful,
-* but WITHOUT ANY WARRANTY; without even the implied warranty of
-* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-* GNU General Public License for more details.
-*
-* You should have received a copy of the GNU General Public License
-* along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
-*/
+/*
+ * particle.h
+ *
+ * Copyright (C) 2001-2023 Daniel Horn, pyramid3d, Stephen G. Tuggy
+ * and other Vega Strike Contributors
+ *
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
+ *
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Vega Strike. If not, see <https://www.gnu.org/licenses/>.
+ */
 
 #ifndef __PARTICLE_H
 #define __PARTICLE_H
@@ -29,6 +28,7 @@
 #include <vector>
 #include <string>
 #include <memory>
+#include <string>
 #include "aligned.h"
 #include "vec.h"
 #include "gfxlib_struct.h"

--- a/engine/src/gldrv/gl_globals.h
+++ b/engine/src/gldrv/gl_globals.h
@@ -1,23 +1,28 @@
 /*
- * Vega Strike
- * Copyright (C) 2001-2002 Daniel Horn & Alan Shieh
+ * gl_globals.h
  *
- * http://vegastrike.sourceforge.net/
+ * Copyright (C) 2001-2023 Daniel Horn, Alan Shieh, pyramid3d,
+ * Stephen G. Tuggy, and other Vega Strike contributors.
  *
- * This program is free software; you can redistribute it and/or
- * modify it under the terms of the GNU General Public License
- * as published by the Free Software Foundation; either version 2
- * of the License, or (at your option) any later version.
+ * https://github.com/vegastrike/Vega-Strike-Engine-Source
  *
- * This program is distributed in the hope that it will be useful,
+ * This file is part of Vega Strike.
+ *
+ * Vega Strike is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Vega Strike is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
- * along with this program; if not, write to the Free Software
- * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ * along with Vega Strike.  If not, see <https://www.gnu.org/licenses/>.
  */
+
+
 #ifndef _GL_GLOBALS_H_
 #define _GL_GLOBALS_H_
 #include <queue>
@@ -142,104 +147,6 @@ struct GFXStats
 #define GL_MAX_CUBE_MAP_TEXTURE_SIZE_EXT 0x851C
 #endif
 
-#if defined (__APPLE__)
-#define glMultiDrawElements_p glMultiDrawElements
-#define glMultiDrawArrays_p glMultiDrawArrays
-#define glMultiTexCoord4fARB_p glMultiTexCoord4fARB
-#define glMultiTexCoord2fARB_p glMultiTexCoord2fARB
-#define glClientActiveTextureARB_p glClientActiveTextureARB
-#define glActiveTextureARB_p glActiveTextureARB
-#define glColorTable_p glColorTable
-#define glCompressedTexImage2D_p glCompressedTexImage2D
-#ifndef OSX_LOWER_THAN_10_4
-#define glGetShaderiv_p glGetShaderiv
-#define glGetProgramiv_p glGetProgramiv
-#define glGetShaderInfoLog_p glGetShaderInfoLog
-#define glGetProgramInfoLog_p glGetProgramInfoLog
-#define glCreateShader_p glCreateShader
-#define glShaderSource_p glShaderSource
-#define glCompileShader_p glCompileShader
-#define glCreateProgram_p glCreateProgram
-#define glAttachShader_p glAttachShader
-#define glLinkProgram_p glLinkProgram
-#define glUseProgram_p glUseProgram
-#define glGetUniformLocation_p glGetUniformLocation
-#define glUniform1f_p glUniform1f
-#define glUniform2f_p glUniform2f
-#define glUniform3f_p glUniform3f
-#define glUniform4f_p glUniform4f
-#define glUniform1i_p glUniform1i
-#define glUniform2i_p glUniform2i
-#define glUniform3i_p glUniform3i
-#define glUniform4i_p glUniform4i
-#define glUniform1fv_p glUniform1fv
-#define glUniform2fv_p glUniform2fv
-#define glUniform3fv_p glUniform3fv
-#define glUniform4fv_p glUniform4fv
-#define glUniform1iv_p glUniform1iv
-#define glUniform2iv_p glUniform2iv
-#define glUniform3iv_p glUniform3iv
-#define glUniform4iv_p glUniform4iv
-
-#define glDeleteShader_p glDeleteShader
-#define glDeleteProgram_p glDeleteProgram
-#else
-#define glGetShaderiv_p( a, b, c )
-#define glGetProgramiv_p( a, b, c )
-#define glGetShaderInfoLog_p( a, b, c, d )
-#define glGetProgramInfoLog_p( a, b, c, d )
-#define glCreateShader_p( a ) 0
-#define glShaderSource_p( a, b, c, d )
-#define glCompileShader_p( a )
-#define glCreateProgram_p() 0
-#define glAttachShader_p( a, b )
-#define glLinkProgram_p( a )
-#define glUseProgram_p( a )
-#define glGetUniformLocation_p( a, b ) 0
-#define glUniform1f_p( a, b )
-#define glUniform2f_p( a, b, c )
-#define glUniform3f_p( a, b, c, d )
-#define glUniform4f_p( a, b, c, d, e )
-#define glUniform1i_p( a, b )
-#define glUniform2i_p( a, b, c )
-#define glUniform3i_p( a, b, c, d )
-#define glUniform4i_p( a, b, c, d, e )
-#define glUniform1fv_p( a, b, c )
-#define glUniform2fv_p( a, b, c )
-#define glUniform3fv_p( a, b, c )
-#define glUniform4fv_p( a, b, c )
-#define glUniform1iv_p( a, b, c )
-#define glUniform2iv_p( a, b, c )
-#define glUniform3iv_p( a, b, c )
-#define glUniform4iv_p( a, b, c )
-
-#define glDeleteProgram_p glIsTexture
-
-#endif
-#if !defined (glLockArraysEXT) || !defined (glUnlockArraysEXT)
-#define NO_COMPILEDVERTEXARRAY_SUPPORT
-#endif
-
-#ifndef NO_COMPILEDVERTEXARRAY_SUPPORT
-#define glLockArraysEXT_p glLockArraysEXT
-#define glUnlockArraysEXT_p glUnlockArraysEXT
-#endif
-
-#if !defined (glDeleteBuffersARB) || !defined (glGenBuffersARB) || !defined (glBindBuffersARB) || !defined (glMapBufferARB) \
-    || !defined (glUnmapBufferARB)
-#define NO_VBO_SUPPORT
-#endif
-
-#ifndef NO_VBO_SUPPORT
-#define glDeleteBuffersARB_p glDeleteBuffersARB
-#define glGenBuffersARB_p glGenBuffersARB
-#define glBindBufferARB_p glBindBufferARB
-#define glBufferDataARB_p glBufferDataARB
-#define glMapBufferARB_p glMapBufferARB
-#define glUnmapBufferARB_p glUnmapBufferARB
-#endif
-
-#else
 extern PFNGLBINDBUFFERARBPROC          glBindBufferARB_p;
 extern PFNGLGENBUFFERSARBPROC          glGenBuffersARB_p;
 extern PFNGLDELETEBUFFERSARBPROC       glDeleteBuffersARB_p;
@@ -256,7 +163,7 @@ extern PFNGLMULTIDRAWARRAYSEXTPROC     glMultiDrawArrays_p;
 extern PFNGLMULTIDRAWELEMENTSEXTPROC   glMultiDrawElements_p;
 extern PFNGLUNLOCKARRAYSEXTPROC        glUnlockArraysEXT_p;
 extern PFNGLCOMPRESSEDTEXIMAGE2DPROC   glCompressedTexImage2D_p;
-extern PFNGLGETSHADERIVPROC glGetShaderiv_p;
+extern PFNGLGETSHADERIVPROC            glGetShaderiv_p;
 extern PFNGLGETPROGRAMIVPROC           glGetProgramiv_p;
 extern PFNGLGETSHADERINFOLOGPROC       glGetShaderInfoLog_p;
 extern PFNGLGETPROGRAMINFOLOGPROC      glGetProgramInfoLog_p;
@@ -265,33 +172,31 @@ extern PFNGLSHADERSOURCEPROC           glShaderSource_p;
 extern PFNGLCOMPILESHADERPROC          glCompileShader_p;
 extern PFNGLCREATEPROGRAMPROC          glCreateProgram_p;
 extern PFNGLATTACHSHADERPROC           glAttachShader_p;
-extern PFNGLLINKPROGRAMPROC glLinkProgram_p;
-extern PFNGLUSEPROGRAMPROC glUseProgram_p;
+extern PFNGLLINKPROGRAMPROC            glLinkProgram_p;
+extern PFNGLUSEPROGRAMPROC             glUseProgram_p;
 extern PFNGLGETUNIFORMLOCATIONPROC     glGetUniformLocation_p;
-extern PFNGLUNIFORM1FPROC     glUniform1f_p;
-extern PFNGLUNIFORM2FPROC     glUniform2f_p;
-extern PFNGLUNIFORM3FPROC     glUniform3f_p;
-extern PFNGLUNIFORM4FPROC     glUniform4f_p;
+extern PFNGLUNIFORM1FPROC              glUniform1f_p;
+extern PFNGLUNIFORM2FPROC              glUniform2f_p;
+extern PFNGLUNIFORM3FPROC              glUniform3f_p;
+extern PFNGLUNIFORM4FPROC              glUniform4f_p;
 
-extern PFNGLUNIFORM1IPROC     glUniform1i_p;
-extern PFNGLUNIFORM2IPROC     glUniform2i_p;
-extern PFNGLUNIFORM3IPROC     glUniform3i_p;
-extern PFNGLUNIFORM4IPROC     glUniform4i_p;
+extern PFNGLUNIFORM1IPROC              glUniform1i_p;
+extern PFNGLUNIFORM2IPROC              glUniform2i_p;
+extern PFNGLUNIFORM3IPROC              glUniform3i_p;
+extern PFNGLUNIFORM4IPROC              glUniform4i_p;
 
-extern PFNGLUNIFORM1FVPROC    glUniform1fv_p;
-extern PFNGLUNIFORM2FVPROC    glUniform2fv_p;
-extern PFNGLUNIFORM3FVPROC    glUniform3fv_p;
-extern PFNGLUNIFORM4FVPROC    glUniform4fv_p;
+extern PFNGLUNIFORM1FVPROC             glUniform1fv_p;
+extern PFNGLUNIFORM2FVPROC             glUniform2fv_p;
+extern PFNGLUNIFORM3FVPROC             glUniform3fv_p;
+extern PFNGLUNIFORM4FVPROC             glUniform4fv_p;
 
-extern PFNGLUNIFORM1IVPROC    glUniform1iv_p;
-extern PFNGLUNIFORM2IVPROC    glUniform2iv_p;
-extern PFNGLUNIFORM3IVPROC    glUniform3iv_p;
-extern PFNGLUNIFORM4IVPROC    glUniform4iv_p;
+extern PFNGLUNIFORM1IVPROC             glUniform1iv_p;
+extern PFNGLUNIFORM2IVPROC             glUniform2iv_p;
+extern PFNGLUNIFORM3IVPROC             glUniform3iv_p;
+extern PFNGLUNIFORM4IVPROC             glUniform4iv_p;
 
-extern PFNGLDELETESHADERPROC  glDeleteShader_p;
-extern PFNGLDELETEPROGRAMPROC glDeleteProgram_p;
-
-#endif /* __APPLE_PANTHER_GCC33_CLI__ */
+extern PFNGLDELETESHADERPROC           glDeleteShader_p;
+extern PFNGLDELETEPROGRAMPROC          glDeleteProgram_p;
 
 //extern int sharedcolortable;
 #ifdef STATS_QUEUE

--- a/engine/src/gldrv/gl_init.cpp
+++ b/engine/src/gldrv/gl_init.cpp
@@ -57,7 +57,11 @@
 #ifndef NOMINMAX
 #define NOMINMAX
 #endif //tells VCC not to generate min/max macros
-#define GL_TEXTURE_CUBE_MAP_SEAMLESS_ARB 0x884F // FIXME May need to actually include the GLEW package for this to work.
+// #if defined (__APPLE__) || defined (MACOSX)
+// #include <GL/glew.h>
+// #else
+#define GL_TEXTURE_CUBE_MAP_SEAMLESS_ARB 0x884F
+// #endif
 #include <windows.h>
 #endif
 #define GL_GLEXT_PROTOTYPES 1

--- a/engine/src/gldrv/gl_init.cpp
+++ b/engine/src/gldrv/gl_init.cpp
@@ -98,7 +98,7 @@
 #include "cg_global.h"
 #endif
 
-NGLBINDBUFFERARBPROC glBindBufferARB_p = nullptr;
+PFNGLBINDBUFFERARBPROC glBindBufferARB_p = nullptr;
 PFNGLGENBUFFERSPROC glGenBuffersARB_p = nullptr;
 PFNGLDELETEBUFFERSARBPROC glDeleteBuffersARB_p = nullptr;
 PFNGLBUFFERDATAARBPROC glBufferDataARB_p = nullptr;

--- a/engine/src/gldrv/gl_init.cpp
+++ b/engine/src/gldrv/gl_init.cpp
@@ -1,9 +1,8 @@
-/**
+/*
  * gl_init.cpp
  *
- * Copyright (C) 2001-2002 Daniel Horn
- * Copyright (C) 2020 pyramid3d, Stephen G. Tuggy, and other Vega Strike
- * contributors
+ * Copyright (C) 2001-2023 Daniel Horn, pyramid3d, Stephen G. Tuggy,
+ * and other Vega Strike contributors.
  *
  * https://github.com/vegastrike/Vega-Strike-Engine-Source
  *
@@ -65,6 +64,7 @@
 #if defined (__APPLE__) || defined (MACOSX)
     #include <OpenGL/gl.h>
     #include <OpenGL/glext.h>
+    #include <dlfcn.h>
 #else
     #include <GL/gl.h>
     #include <GL/glext.h>
@@ -78,7 +78,7 @@
 //typedef void (APIENTRY * PFNGLLOCKARRAYSEXTPROC) (GLint first, GLsizei count);
 //typedef void (APIENTRY * PFNGLUNLOCKARRAYSEXTPROC) (void);
 
-#if !defined (__APPLE__) && !defined (MACOSX) && !defined (WIN32)  && !defined (__HAIKU__)
+#if !defined (WIN32) && !defined (__HAIKU__) && !defined (__APPLE__) && !defined (MACOSX)
     # define GLX_GLXEXT_PROTOTYPES 1
     # define GLX_GLXEXT_LEGACY 1
     # include <GL/glx.h>
@@ -94,74 +94,76 @@
 #include "cg_global.h"
 #endif
 
-#if !defined (__APPLE__)
+NGLBINDBUFFERARBPROC glBindBufferARB_p = nullptr;
+PFNGLGENBUFFERSPROC glGenBuffersARB_p = nullptr;
+PFNGLDELETEBUFFERSARBPROC glDeleteBuffersARB_p = nullptr;
+PFNGLBUFFERDATAARBPROC glBufferDataARB_p = nullptr;
+PFNGLMAPBUFFERARBPROC glMapBufferARB_p = nullptr;
+PFNGLUNMAPBUFFERARBPROC glUnmapBufferARB_p = nullptr;
 
-PFNGLBINDBUFFERARBPROC    glBindBufferARB_p    = 0;
-PFNGLGENBUFFERSARBPROC    glGenBuffersARB_p    = 0;
-PFNGLDELETEBUFFERSARBPROC glDeleteBuffersARB_p = 0;
-PFNGLBUFFERDATAARBPROC    glBufferDataARB_p    = 0;
-PFNGLMAPBUFFERARBPROC     glMapBufferARB_p     = 0;
-PFNGLUNMAPBUFFERARBPROC   glUnmapBufferARB_p   = 0;
+PFNGLCLIENTACTIVETEXTUREARBPROC glClientActiveTextureARB_p = nullptr;
+PFNGLCLIENTACTIVETEXTUREARBPROC glActiveTextureARB_p = nullptr;
+PFNGLCOLORTABLEEXTPROC glColorTable_p = nullptr;
+PFNGLMULTITEXCOORD2FARBPROC glMultiTexCoord2fARB_p = nullptr;
+PFNGLMULTITEXCOORD4FARBPROC glMultiTexCoord4fARB_p = nullptr;
 
-PFNGLCLIENTACTIVETEXTUREARBPROC glClientActiveTextureARB_p = 0;
-PFNGLCLIENTACTIVETEXTUREARBPROC glActiveTextureARB_p = 0;
-PFNGLCOLORTABLEEXTPROC glColorTable_p = 0;
-PFNGLMULTITEXCOORD2FARBPROC     glMultiTexCoord2fARB_p     = 0;
-PFNGLMULTITEXCOORD4FARBPROC     glMultiTexCoord4fARB_p     = 0;
+PFNGLLOCKARRAYSEXTPROC glLockArraysEXT_p = nullptr;
+PFNGLUNLOCKARRAYSEXTPROC glUnlockArraysEXT_p = nullptr;
+PFNGLCOMPRESSEDTEXIMAGE2DPROC glCompressedTexImage2D_p = nullptr;
+PFNGLMULTIDRAWARRAYSEXTPROC glMultiDrawArrays_p = nullptr;
+PFNGLMULTIDRAWELEMENTSEXTPROC glMultiDrawElements_p = nullptr;
 
-PFNGLLOCKARRAYSEXTPROC        glLockArraysEXT_p        = 0;
-PFNGLUNLOCKARRAYSEXTPROC      glUnlockArraysEXT_p      = 0;
-PFNGLCOMPRESSEDTEXIMAGE2DPROC glCompressedTexImage2D_p = 0;
-PFNGLMULTIDRAWARRAYSEXTPROC   glMultiDrawArrays_p      = 0;
-PFNGLMULTIDRAWELEMENTSEXTPROC glMultiDrawElements_p    = 0;
+PFNGLGETSHADERIVPROC glGetShaderiv_p = nullptr;
+PFNGLGETPROGRAMIVPROC glGetProgramiv_p = nullptr;
+PFNGLGETSHADERINFOLOGPROC glGetShaderInfoLog_p = nullptr;
+PFNGLGETPROGRAMINFOLOGPROC glGetProgramInfoLog_p = nullptr;
+PFNGLCREATESHADERPROC glCreateShader_p = nullptr;
+PFNGLSHADERSOURCEPROC glShaderSource_p = nullptr;
+PFNGLCOMPILESHADERPROC glCompileShader_p = nullptr;
+PFNGLCREATEPROGRAMPROC glCreateProgram_p = nullptr;
+PFNGLATTACHSHADERPROC glAttachShader_p = nullptr;
+PFNGLLINKPROGRAMPROC glLinkProgram_p = nullptr;
+PFNGLUSEPROGRAMPROC glUseProgram_p = nullptr;
+PFNGLGETUNIFORMLOCATIONPROC glGetUniformLocation_p = nullptr;
+PFNGLUNIFORM1FPROC glUniform1f_p = nullptr;
+PFNGLUNIFORM2FPROC glUniform2f_p = nullptr;
+PFNGLUNIFORM3FPROC glUniform3f_p = nullptr;
+PFNGLUNIFORM4FPROC glUniform4f_p = nullptr;
 
-PFNGLGETSHADERIVPROC        glGetShaderiv_p        = 0;
-PFNGLGETPROGRAMIVPROC       glGetProgramiv_p       = 0;
-PFNGLGETSHADERINFOLOGPROC   glGetShaderInfoLog_p   = 0;
-PFNGLGETPROGRAMINFOLOGPROC  glGetProgramInfoLog_p  = 0;
-PFNGLCREATESHADERPROC       glCreateShader_p       = 0;
-PFNGLSHADERSOURCEPROC       glShaderSource_p       = 0;
-PFNGLCOMPILESHADERPROC      glCompileShader_p      = 0;
-PFNGLCREATEPROGRAMPROC      glCreateProgram_p      = 0;
-PFNGLATTACHSHADERPROC       glAttachShader_p       = 0;
-PFNGLLINKPROGRAMPROC        glLinkProgram_p        = 0;
-PFNGLUSEPROGRAMPROC glUseProgram_p = 0;
-PFNGLGETUNIFORMLOCATIONPROC glGetUniformLocation_p = 0;
-PFNGLUNIFORM1FPROC     glUniform1f_p     = 0;
-PFNGLUNIFORM2FPROC     glUniform2f_p     = 0;
-PFNGLUNIFORM3FPROC     glUniform3f_p     = 0;
-PFNGLUNIFORM4FPROC     glUniform4f_p     = 0;
+PFNGLUNIFORM1IPROC glUniform1i_p = nullptr;
+PFNGLUNIFORM2IPROC glUniform2i_p = nullptr;
+PFNGLUNIFORM3IPROC glUniform3i_p = nullptr;
+PFNGLUNIFORM4IPROC glUniform4i_p = nullptr;
 
-PFNGLUNIFORM1IPROC     glUniform1i_p     = 0;
-PFNGLUNIFORM2IPROC     glUniform2i_p     = 0;
-PFNGLUNIFORM3IPROC     glUniform3i_p     = 0;
-PFNGLUNIFORM4IPROC     glUniform4i_p     = 0;
+PFNGLUNIFORM1FVPROC glUniform1fv_p = nullptr;
+PFNGLUNIFORM2FVPROC glUniform2fv_p = nullptr;
+PFNGLUNIFORM3FVPROC glUniform3fv_p = nullptr;
+PFNGLUNIFORM4FVPROC glUniform4fv_p = nullptr;
 
-PFNGLUNIFORM1FVPROC    glUniform1fv_p    = 0;
-PFNGLUNIFORM2FVPROC    glUniform2fv_p    = 0;
-PFNGLUNIFORM3FVPROC    glUniform3fv_p    = 0;
-PFNGLUNIFORM4FVPROC    glUniform4fv_p    = 0;
+PFNGLUNIFORM1IVPROC glUniform1iv_p = nullptr;
+PFNGLUNIFORM2IVPROC glUniform2iv_p = nullptr;
+PFNGLUNIFORM3IVPROC glUniform3iv_p = nullptr;
+PFNGLUNIFORM4IVPROC glUniform4iv_p = nullptr;
 
-PFNGLUNIFORM1IVPROC    glUniform1iv_p    = 0;
-PFNGLUNIFORM2IVPROC    glUniform2iv_p    = 0;
-PFNGLUNIFORM3IVPROC    glUniform3iv_p    = 0;
-PFNGLUNIFORM4IVPROC    glUniform4iv_p    = 0;
-
-PFNGLDELETESHADERPROC  glDeleteShader_p  = 0;
-PFNGLDELETEPROGRAMPROC glDeleteProgram_p = 0;
-
-#endif /* __APPLE_PANTHER_GCC33_CLI__ */
+PFNGLDELETESHADERPROC glDeleteShader_p = nullptr;
+PFNGLDELETEPROGRAMPROC glDeleteProgram_p = nullptr;
 
 typedef void ( *(*get_gl_proc_fptr_t)(const GLubyte*) )();
 #ifdef _WIN32
-typedef char*GET_GL_PTR_TYP;
-#define GET_GL_PROC wglGetProcAddress
+    typedef char * GET_GL_PTR_TYP;
+    #define GET_GL_PROC wglGetProcAddress
 #else
 	#if defined(__HAIKU__)
 	    typedef char * GET_GL_PTR_TYP;
 		#define GET_GL_PROC glutGetProcAddress
+    #elif defined (__MACOSX__)
+        typedef const char * GET_GL_PTR_TYP;
+        void * vega_dlsym(GET_GL_PTR_TYP function_name) {
+            return dlsym(RTLD_SELF, function_name);
+        }
+        #define GET_GL_PROC vega_dlsym
 	#else
-		typedef GLubyte*GET_GL_PTR_TYP;
+		typedef GLubyte *GET_GL_PTR_TYP;
 		#define GET_GL_PROC glXGetProcAddressARB
 	#endif
 #endif
@@ -173,7 +175,7 @@ CG_Cloak *cloak_cg = new CG_Cloak();
 /* CENTRY */
 int vsExtensionSupported( const char *extension )
 {
-    static const GLubyte *extensions = NULL;
+    static const GLubyte *extensions = nullptr;
     const GLubyte *start;
     GLubyte *where, *terminator;
 
@@ -256,9 +258,12 @@ void init_opengl_extensions()
         glLockArraysEXT_p   = &glLockArraysEXT;
         glUnlockArraysEXT_p = &glUnlockArraysEXT;
 #else
-        glLockArraysEXT_p   = 0;
-        glUnlockArraysEXT_p = 0;
+        glLockArraysEXT_p   = nullptr;
+        glUnlockArraysEXT_p = nullptr;
 #endif
+#else
+        glLockArraysEXT_p = (PFNGLLOCKARRAYSEXTPROC)(dlsym(RTLD_SELF, "glLockArraysEXT"));
+        glUnlockArraysEXT_p = (PFNGLUNLOCKARRAYSEXTPROC)(dlsym(RTLD_SELF, "glUnlockArraysEXT"));
 #endif /*__APPLE_PANTHER_GCC33_CLI__*/
 #else
         glLockArraysEXT_p   = (PFNGLLOCKARRAYSEXTPROC)
@@ -269,40 +274,39 @@ void init_opengl_extensions()
         BOOST_LOG_TRIVIAL(trace) << "OpenGL::GL_EXT_compiled_vertex_array supported";
     } else {
 #ifdef __APPLE__
-#ifndef __APPLE_PANTHER_GCC33_CLI__
-        glLockArraysEXT_p   = 0;
-        glUnlockArraysEXT_p = 0;
-#endif /*__APPLE_PANTHER_GCC33_CLI__*/
+        glLockArraysEXT_p   = nullptr;
+        glUnlockArraysEXT_p = nullptr;
 #endif
         BOOST_LOG_TRIVIAL(debug) << "OpenGL::GL_EXT_compiled_vertex_array unsupported";
     }
 #endif
-#ifndef __APPLE__
-    if ( vsExtensionSupported( "GL_EXT_multi_draw_arrays" ) ) {
+#if defined (__MACOSX__)
+    if (vsExtensionSupported("GL_EXT_multi_draw_arrays")) {
         glMultiDrawArrays_p   = (PFNGLMULTIDRAWARRAYSEXTPROC)
                                 GET_GL_PROC( (GET_GL_PTR_TYP) "glMultiDrawArraysEXT" );
         glMultiDrawElements_p = (PFNGLMULTIDRAWELEMENTSEXTPROC)
                                 GET_GL_PROC( (GET_GL_PTR_TYP) "glMultiDrawElementsEXT" );
         BOOST_LOG_TRIVIAL(trace) << "OpenGL::GL_EXT_multi_draw_arrays supported";
     } else {
-        glMultiDrawArrays_p   = 0;
-        glMultiDrawElements_p = 0;
+        glMultiDrawArrays_p   = nullptr;
+        glMultiDrawElements_p = nullptr;
         BOOST_LOG_TRIVIAL(debug) << "OpenGL::GL_EXT_multi_draw_arrays unsupported";
     }
+#else
+    if (vsExtensionSupported("GL_EXT_multi_draw_arrays")) {
+        glMultiDrawArrays_p = (PFNGLMULTIDRAWARRAYSEXTPROC)
+                GET_GL_PROC((GET_GL_PTR_TYP) "glMultiDrawArraysEXT");
+        glMultiDrawElements_p = (PFNGLMULTIDRAWELEMENTSEXTPROC)
+                GET_GL_PROC((GET_GL_PTR_TYP) "glMultiDrawElementsEXT");
+        VS_LOG(trace, "OpenGL::GL_EXT_multi_draw_arrays supported");
+    } else {
+        glMultiDrawArrays_p = nullptr;
+        glMultiDrawElements_p = nullptr;
 #endif
 
-#ifdef __APPLE__
-#ifndef __APPLE__
-    glColorTable_p = glColorTableEXT;
-    glMultiTexCoord2fARB_p     = glMultiTexCoord2fARB;
-    glMultiTexCoord4fARB_p     = glMultiTexCoord4fARB;
-    glClientActiveTextureARB_p = glClientActiveTextureARB;
-    glActiveTextureARB_p = glActiveTextureARB;
-#endif /*__APPLE_PANTHER_GCC33_CLI__*/
-#else
 #ifndef NO_VBO_SUPPORT
     glBindBufferARB_p    = (PFNGLBINDBUFFERARBPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glBindBuffer" );
-    glGenBuffersARB_p    = (PFNGLGENBUFFERSARBPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glGenBuffers" );
+    glGenBuffersARB_p    = (PFNGLGENBUFFERSPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glGenBuffers" );
     glDeleteBuffersARB_p = (PFNGLDELETEBUFFERSARBPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glDeleteBuffers" );
     glBufferDataARB_p    = (PFNGLBUFFERDATAARBPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glBufferData" );
     glMapBufferARB_p     = (PFNGLMAPBUFFERARBPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glMapBuffer" );
@@ -385,7 +389,6 @@ void init_opengl_extensions()
     if (!glDeleteProgram_p)
         glDeleteProgram_p = (PFNGLDELETEPROGRAMPROC) GET_GL_PROC( (GET_GL_PTR_TYP) "glDeleteProgram" );
     //fixme
-#endif
 
 #ifdef GL_FOG_DISTANCE_MODE_NV
     if ( vsExtensionSupported( "GL_NV_fog_distance" ) ) {

--- a/engine/src/gldrv/gl_init.cpp
+++ b/engine/src/gldrv/gl_init.cpp
@@ -302,10 +302,11 @@ void init_opengl_extensions()
                 GET_GL_PROC((GET_GL_PTR_TYP) "glMultiDrawArraysEXT");
         glMultiDrawElements_p = (PFNGLMULTIDRAWELEMENTSEXTPROC)
                 GET_GL_PROC((GET_GL_PTR_TYP) "glMultiDrawElementsEXT");
-        VS_LOG(trace, "OpenGL::GL_EXT_multi_draw_arrays supported");
+        BOOST_LOG_TRIVIAL(trace) << "OpenGL::GL_EXT_multi_draw_arrays supported";
     } else {
         glMultiDrawArrays_p = nullptr;
         glMultiDrawElements_p = nullptr;
+    }
 #endif
 
 #ifndef NO_VBO_SUPPORT

--- a/engine/src/gldrv/gl_program.cpp
+++ b/engine/src/gldrv/gl_program.cpp
@@ -1,9 +1,8 @@
-/**
+/*
  * gl_program.cpp
  *
- * Copyright (C) Daniel Horn
- * Copyright (C) 2020 pyramid3d, Stephen G. Tuggy, and other Vega Strike
- * contributors
+ * Copyright (C) 2001-2023 Daniel Horn, pyramid3d, Stephen G. Tuggy,
+ * and other Vega Strike contributors.
  *
  * https://github.com/vegastrike/Vega-Strike-Engine-Source
  *

--- a/engine/src/vsfilesystem.cpp
+++ b/engine/src/vsfilesystem.cpp
@@ -1,8 +1,7 @@
-/**
+/*
  * vsfilesystem.cpp
  *
- * Copyright (C) Daniel Horn
- * Copyright (C) 2020-2021 pyramid3d, Nachum Barcohen, Roy Falk,
+ * Copyright (C) 2001-2023 Daniel Horn, pyramid3d, Nachum Barcohen, Roy Falk,
  * Stephen G. Tuggy, and other Vega Strike contributors
  *
  * https://github.com/vegastrike/Vega-Strike-Engine-Source
@@ -100,11 +99,7 @@ extern string GetUnitDir( string filename );
 
 string selectcurrentdir;
 
-#if defined (__APPLE__)
-int selectdirs( struct dirent *entry )
-#else
 int selectdirs( const struct dirent * entry )
-#endif
 {
     if (string( entry->d_name ) == "." && string( entry->d_name ) == "..")
         return 0;

